skbio/stats/ordination/_principal_coordinate_analysis.py

```diff
@@ -130,15 +153,170 @@ def pcoa(distance_matrix):
+           np.array
+               Array of eigenvectors, each with num_dimensions_out length.
+           np.array
+               Array of eigenvalues, a total number of num_dimensions_out.
+
+           Notes
+           -----
+           The algorithm is based on 'An Algorithm for the Principal
+           Component analysis of Large Data Sets'
+           by N. Halko, P.G. Martinsson, Y. Shkolnisky, and M. Tygert.
+           Original Paper: https://arxiv.org/abs/1007.5510
+
+           Ported from reference MATLAB implementation: https://goo.gl/JkcxQ2
+           """
+
+    m, n = centered_distance_matrix.shape
+
+    # Note: a (conjugate) transpose is removed for performance, since we
+    # only expect square matrices.
+    if m != n:
+        raise ValueError('FSVD expects square distance matrix')
+
+    k = dimension + 2
+
+    # Form a real nxl matrix G whose entries are independent,
+    # identically distributed
+    # Gaussian random variables of
+    # zero mean and unit variance
+    G = standard_normal(size=(n, k))
+
+    if use_power_method:
+        # use only the given exponent
+        H = dot(centered_distance_matrix, G)
+
+        for x in range(2, num_levels + 2):
+            # enhance decay of singular values
+            # note: distance_matrix is no longer transposed, saves work
+            # since we're expecting symmetric, square matrices anyway
+            # (Daniel McDonald's changes)
+            H = dot(centered_distance_matrix, dot(centered_distance_matrix, H))
+
+    else:
+        # compute the m x l matrices H^{(0)}, ..., H^{(i)}
+        # Note that this is done implicitly in each iteration below.
+        H = dot(centered_distance_matrix, G)
+        # Again, removed transpose: dot(distance_matrix.transpose(), H)
+        # to enhance performance
+        H = hstack(
+            (H,
+             dot(centered_distance_matrix, dot(centered_distance_matrix, H))))
+        for x in range(3, num_levels + 2):
+            # Removed this transpose: dot(distance_matrix.transpose(), H)
+            tmp = dot(centered_distance_matrix,
+                      dot(centered_distance_matrix, H))
+
+            # Removed this transpose: dot(distance_matrix.transpose(), tmp)
+            H = hstack(
+                (H, dot(centered_distance_matrix,
+                        dot(centered_distance_matrix, tmp))))
+
+    # Using the pivoted QR-decomposition, form a real m * ((i+1)l) matrix Q
+    # whose columns are orthonormal, s.t. there exists a real
+    # ((i+1)l) * ((i+1)l) matrix R for which H = QR
+    Q, R = qr(H)
+
+    # Compute the n * ((i+1)l) product matrix T = A^T Q
+    # Removed transpose of distance_matrix for performance
+    T = dot(centered_distance_matrix, Q)  # step 3
+
+    # Form an SVD of T
+    Vt, St, W = svd(T, full_matrices=False)
+    W = W.transpose()
+
+    # Compute the m * ((i+1)l) product matrix
+    Ut = dot(Q, W)
+
+    # V_fsvd = Vt[:, :num_dimensions_out] # unused
```

Comment:
Can we remove comment? If not my guess is that num_dimensions_out -> dimension, right?